odoo/http.py

```diff
@@ -950,6 +955,40 @@ def vacuum(self, max_lifetime=SESSION_LIFETIME):
                 if os.path.getmtime(path) < threshold:
                     os.unlink(path)
 
+    def generate_key(self, salt=None):
+        # About the generated key:
+        # ------------------------
+        # Now, the generated key is case sensitive (base64) and the length is 84 chars.
+        # In the worst-case scenario, i.e. in an insensitive filesystem (NTFS for example)
+        # taking into account the proportion of characters in the pool and a length
+        # of 42 (stored part in the database), the entropy for the base64 generated key
+        # is 217.875 bits which is better than the previous implementation
+        # (160 bits, entropy of hex with a length of 40).
+        # The risk of collision is negligible in practice.
+        # Formulas:
+        #   - L: length of generated word
+        #   - p_char: probability of obtaining the character in the pool
+        #   - n: size of the pool
+        #   - k: number of generated word
+        #   Entropy = - L * sum(p_char * log2(p_char))
+        #   Collision ~= (1 - exp((-k * (k - 1)) / (2 * (n**L))))
+        key = str(time.time()).encode() + os.urandom(64)
+        hash_key = sha512(key).digest()[:-1]  # prevent base64 padding
+        return base64.urlsafe_b64encode(hash_key).decode('utf-8')
+
+    def is_valid_key(self, key):
+        return _base64_urlsafe_re.match(key) is not None
+
+    def delete_from_identifiers(self, identifiers):
+        files_to_unlink = []
+        for identifier in identifiers:
+            normalized_path = os.path.normpath(os.path.join(self.path, identifier[:2], identifier + '*'))
```

Comment:
should we check that we have a properly formed identifier? e.g. could I bruteforce all 2 letters identifiers to delete every sessions

Reply:
A check for the identifier length should be added.

Reply:
Thanks for this remark !